src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1765,6 +1780,7 @@ private void RemoveStream(Http2Stream http2Stream)
             }
 
             _concurrentStreams.AdjustCredit(1);
+            _canAddNewStream = true;
```

Comment:
Could you please clarify more the suspicious case?
As far as I see it the code above is invoked **only** if there is a waiter, but if `MaxHttp2ConnectionsPerServer` is >1 `_http2Connections` > `MaxHttp2ConnectionsPerServer` and consequently [_pool.ThrowOnStreamLimitReached](https://github.com/dotnet/runtime/blob/d97e37ef01eee58e6d3eb08342ee626d2abf8473/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1169) is true, then only `TryRequestCreditNoWait` gets invoked that means `_waiterList` is always null. Once, the maximum of HTTP/2 connections reached, `_pool.ThrowOnStreamLimitReached` gets `false` and now `_canAddNewStream` is not anymore taken into account.

Reply:
I get it now. Even if `_canAddNewStream` is being set to `true` here while the actual credit might still be 0. It will be handled in `SendHeadersAsync` by throwing an exception and retrying on a different connection.

So it's rather `shouldTryAddNewStreamButMightMakeYouRetryOnDifferentConnection`, though that's quite mouthful :smile: